### PR TITLE
Add vault integration

### PR DESCRIPTION
**Pull request by Jata2k**
Add a vault provider and swap variables for vault values

Resolves: DVOP-1455

### CODE REVIEW

File: groups/data-lake/main.tf [hunk 1]
```diff
@@ -14,6 +10,24 @@ data "aws_security_group" "mongo_db" {
   }
 }
 
+data "vault_generic_secret" "secrets" {
+  path = "applications/${var.aws_profile}/data-lake"
```
Comment (marcransome) [inline groups/data-lake/main.tf]: Can introduce a `service` variable with a default value of `data-lake` to make this more consistent with other projects?

```
variable "service" {
  type        = string
  description = "The service name to be used when creating AWS resources"
  default     = "data-lake"
}
```

Comment (marcransome) [reply groups/data-lake/main.tf]: It's also worth considering whether we will only ever need one compound configuration entry in Vault related to this service, or whether to create an additional level in the path to allow for related entries in future. We've adopted the latter approach in a few places, where the 'service' name is just another subdivision in the path, for example: https://github.com/companieshouse/sftp-gateway-terraform/blob/4ca90cbb4b5db5306e577cc7f8cf875ddc8a5d20/groups/sftp-gateway/sftp-gateway/transfer.tf#L2

File: groups/data-lake/main.tf [hunk 2]
```diff
@@ -14,6 +10,24 @@ data "aws_security_group" "mongo_db" {
...
+}
+
+# ------------------------------------------------------------------------------
+# Locals
+# ------------------------------------------------------------------------------
+locals {
+  database_password = data.vault_generic_secret.secrets.data.database_password
+  database_username = data.vault_generic_secret.secrets.data.database_username
+  mongo_export_db_url = data.vault_generic_secret.secrets.data.mongo_export_db_url
+  mongo_export_s3_path = data.vault_generic_secret.secrets.data.mongo_export_s3_path
```
Comment (marcransome) [inline groups/data-lake/main.tf]: Can we align the `=` character in each variable declaration here so make the code more consistent with other parts of the code?

File: groups/data-lake/main.tf [hunk 3]
```diff
@@ -14,6 +10,24 @@ data "aws_security_group" "mongo_db" {
   }
 }
 
+data "vault_generic_secret" "secrets" {
+  path = "applications/${var.aws_profile}/data-lake"
+}
+
+# ------------------------------------------------------------------------------
+# Locals
+# ------------------------------------------------------------------------------
```
Comment (marcransome) [inline groups/data-lake/main.tf]: Can we drop the separator comments like this from the file? The 'Locals' comment here is followed by a `locals` block and the `Resources` comment later in the file is followed immediately by a `resource` so should be self explanatory, and we always have the option to split files code into separate files if some more separation is needed.

File: groups/data-lake/variables.tf [hunk 4]
```diff
@@ -1,22 +1,8 @@
...
+  default = "development-eu-west-2"
+  type = string
 }
 
-variable "mongo_export_s3_path" {
-  type        = string
-  default     = "DMS-Output/lambda"
-  description = "The S3 bucket prefix for the export Lambda; will be prefixed with the bucket name automatically"
+variable "region" {
+    type = string
```
Comment (marcransome) [inline groups/data-lake/variables.tf]: As above, can we add a suitable description here? For example, this is used in other projects:

```
description = "The AWS region in which resources will be created"
```

File: groups/data-lake/variables.tf [hunk 5]
```diff
@@ -1,22 +1,8 @@
...
-  type        = string
-  description = "The Redshift database username"
-}
-
-variable "mongo_export_db_url" {
-  type        = string
-  description = "The MongoDB URL for the export Lambda"
+variable "aws_profile" {
+  default = "development-eu-west-2"
+  type = string
```
Comment (marcransome) [inline groups/data-lake/variables.tf]: Can we document all the variables for completeness? Here's what we use in other projects:

```
description = "The AWS profile name; used as a prefix for Vault secrets"
```

File: groups/data-lake/vault/token [hunk 6]
```diff
@@ -0,0 +1,5 @@
+provider "vault" {
+    # Credentials read from:
```
Comment (marcransome) [inline groups/data-lake/vault/token]: Can we clarify this slightly: `Credentials read from the environment variables`

File: groups/data-lake/vault/approle [hunk 7]
```diff
@@ -0,0 +1,20 @@
+variable "hashicorp_vault_role_id" {
+  description = "Your Hashicorp Vault role id"
+  type = string
+}
+
+variable "hashicorp_vault_secret_id" {
+  description = "Your Hashicorp Vault secret id"
+  type = string
+}
```
Comment (marcransome) [inline groups/data-lake/vault/approle]: Can we make the purpose of these variables a little clearer? I'm thinking something along the lines of:

`hashicorp_vault_role_id `:
 `The role identifier used when retrieving configuration from Hashicorp Vault`

`hashicorp_vault_secret_id`:
 `The secret identifier used when retrieving configuration from Hashicorp Vault`

If we do make changes, could we also update the variable descriptions in `groups/data-lake/vault/userpass` to be consistently worded?

**Review by marcransome — CHANGES_REQUESTED** [review]

File: groups/data-lake/main.tf (hunk 1 above)
Comment (Jata2k) [reply groups/data-lake/main.tf]: I agree this should use `service` as per other repositories. As for adding another layer within the hierarchy, as we've done to support environments elsewhere, I'm not sure it's required at the moment as I believe this is an account level project. Were we to introduce environment level config we'd need to revise other aspects of the code anyway. I'll add the service variable now...

File: groups/data-lake/main.tf (hunk 3 above)
Comment (Jata2k) [reply groups/data-lake/main.tf]: Fair point. It was more for visual separation given how large the initial implementation of `main.tf`

**Review by marcransome — APPROVED** [review]

**Pull request merged**